openedx/core/djangoapps/session_inactivity_timeout/tests/test_middleware.py

```diff
@@ -0,0 +1,219 @@
+        "",    # Empty string timestamp in session
+    )
+    @override_settings(SESSION_INACTIVITY_TIMEOUT_IN_SECONDS=300)
+    @patch("openedx.core.djangoapps.session_inactivity_timeout.middleware.datetime")
+    @patch("openedx.core.djangoapps.session_inactivity_timeout.middleware.log")
+    @patch(
+        "openedx.core.djangoapps.session_inactivity_timeout.middleware.monitoring_utils"
+    )
+    def test_process_request_first_visit_sets_timestamp(
+        self, timestamp_value, mock_monitoring, mock_log, mock_datetime
+    ):
+        if timestamp_value is not None:
+            self.request.session[LAST_TOUCH_KEYNAME] = timestamp_value
+        # else: leave the session without the timestamp key (None case)
+
+        mock_now = datetime(2025, 6, 16, 12, 0, 0)
+        mock_datetime.utcnow.return_value = mock_now
+
+        response = self.middleware.process_request(self.request)  # lint-amnesty, pylint: disable=assignment-from-none
+
+        assert response is None
+        assert self.request.session[LAST_TOUCH_KEYNAME] == mock_now.isoformat()
+        mock_log.debug.assert_called_once_with("No previous activity timestamp found (first login)")
+
+        mock_monitoring.set_custom_attribute.assert_has_calls([
+            call("session_inactivity.first_login", True),
+            call("session_inactivity.proceed_with_period_save", True),
+        ], any_order=True)
+
+    @ddt.data(
+        "invalid-timestamp",     # Invalid ISO format
+        "2025-13-01T12:00:00",  # Invalid date
+    )
+    @override_settings(SESSION_INACTIVITY_TIMEOUT_IN_SECONDS=300)
+    @patch("openedx.core.djangoapps.session_inactivity_timeout.middleware.log")
+    @patch("openedx.core.djangoapps.session_inactivity_timeout.middleware.monitoring_utils")
+    def test_process_request_invalid_timestamp_handling(
+        self, invalid_timestamp, mock_monitoring, mock_log
+    ):
+        self.request.session[LAST_TOUCH_KEYNAME] = invalid_timestamp
+
+        # The middleware should raise an exception when it tries to parse the invalid timestamp
+        # in the save delay logic (after already catching it once in the timeout logic)
+        with self.assertRaises(ValueError):
+            self.middleware.process_request(self.request)
+
+        mock_log.warning.assert_called_once()
+
+        mock_monitoring.set_custom_attribute.assert_any_call("session_inactivity.last_touch_error", ANY)
+        mock_monitoring.record_exception.assert_called_once()
+
+    @ddt.data(
+        # (timeout_seconds, seconds_elapsed, should_logout)
+        # Test timeout behavior including boundary conditions
+        (300, 240, False),   # 240 sec < 300 sec timeout, no logout
+        (300, 299, False),   # 299 sec < 300 sec timeout, no logout
+        (300, 300, False),   # 300 sec = 300 sec timeout, no logout (not exceeded)
+        (300, 301, True),    # 301 sec > 300 sec timeout, logout occurs
+        (300, 360, True),    # 360 sec > 300 sec timeout, logout occurs
```

Comment:
@ttak-apphelix: I would have thought you'd apply the same removal of non-boundary tests to all tests. I'm going to commit this change now.
```suggestion
        (300, 299, False),   # 299 sec < 300 sec timeout, no logout
        (300, 300, False),   # 300 sec = 300 sec timeout, no logout (not exceeded)
        (300, 301, True),    # 301 sec > 300 sec timeout, logout occurs
```